### PR TITLE
Terra-react-svg-loader review

### PR DESCRIPTION
**Pull request by Matt-Butler**
### Summary
The terra-react-svg-loader takes an svg file as input and transforms it into a react component.

1) Reads svg
2) Runs svg through [svgo](https://github.com/svg/svgo) optimizer (removes unwanted attributes, comments, etc.)
3) Transforms all of the attributes on the svg and its children to jsx attributes.
4) Adds props back to a react component template. The component template includes icon styles. 

@cerner/terra

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,11 +1,6 @@
 # Terra React Svg Loader
```
Comment (mhemesath) [inline README.md]: this document should explain how to hook up the loader and have documented examples.

File: lib/icon.scss [hunk 2]
```diff
@@ -0,0 +1,36 @@
+// Icon Component
```
Comment (mhemesath) [inline lib/icon.scss]: it feels weird for this code to be in this repo. I'd more expect this to reside terra-Icon. Otherwise now I have to declare this project as a depedency when really it sould justs be a build dependency.

File: src/template.txt [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
```
Comment (mhemesath) [inline src/template.txt]: I'd actually prefer that the majority of the logic in this file moves to terra-Icon.  TerraIcon would then just be a SVG icon wrapper that takes child content. This has a few benefits in being easier to read and creates an explicit dependency on the SCSS which should probably be moved to that project anyway. With that impl. this could would really look like this.

```js
import Icon from 'terra-icons';

const Icon = ({...customProps}) => {
  
return <Icon {...customProps><%= icon.children %=></Icon>;
```

The other benefit of this way is that if I need to load a SVG outside of webpack (example client configured) then I can still use TerraIcon and dangerously set the InnerHTML to that icon source.. allowing me to use TerraIcon outside of webpack.

File: README.md (hunk 1 above)
Comment (Matt-Butler) [inline README.md]: +1 I have the documentation story to do right after this... but people will probably want to consume this right away, so I will add docs on doing that.

File: lib/icon.scss (hunk 2 above)
Comment (Matt-Butler) [inline lib/icon.scss]: This was my motivation to moving the styling into the loader.

1) Where should the examples for icon live?
* They should live in terra-icon. terra-icon requires terra-react-svg-loader in order to display the examples. terra-icon needs terra-react-svg-loader as a devDependency to load the examples. 

2) Where should the scss live? 
* The scss should be imported by the react component to be consistent with our other components. 

3) Can the react-component live in the terra-icon repo?  If the react component lives in the terra-icon repo,  then terra-icon needs to be a dependency for terra-react-svg-loader. 

This turns into a weird, almost circular dependency.
* terra-icon has devDependency on terra-react-svg-loader.
* terra-react-svg-loader has dependency on terra-icon. 

Does it make sense to make this change?

Comment (mhemesath) [inline lib/icon.scss]: > terra-icon needs to be a dependency for terra-react-svg-loader

No it doesnt, its a dev dependency. The loader doesn't actually need the SCSS to operate. You *only* need it in dev if you intend to run live examples via svg loader doc. However, I don't think thats necessary. So while the loader generates code that depends on terra-Icon.. it in of itself has no dependency on it.

Comment (Matt-Butler) [inline lib/icon.scss]: Assuming the react/scss is the terra-icon repository, how does terra-react-svg-loader component template import the react component from terra-icon?

Comment (mhemesath) [inline lib/icon.scss]: Consumer app has the following:

*  Dependency on terra-icon
*  Build dependency on terra-svg-loader

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import svgToIcon from './svgToIcon';
+import optimizeSvg from './optimizeSvg';
```
Comment (mhemesath) [inline src/index.js]: are you using this?

Comment (Matt-Butler) [inline src/index.js]: yes, that is the first step:https://github.com/cerner/terra-react-svg-loader/pull/1/files#diff-1fdf421c05c1140f6d71444ea2b27638R10

File: src/template.txt (hunk 3 above)
Comment (Matt-Butler) [inline src/template.txt]: I'm going to move the react component over to terra-Icon today.  

I was originally under the assumption that I could not have 
```
terra-react-svg-loader >(dependency)> terra-icon
terra-icon >(devDependency)> terra-react-svg-loader 
```
but it is possible.

File: src/index.js (hunk 4 above)
Comment (mhemesath) [inline src/index.js]: oh duh hahaha

File: src/template.txt [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+import Icon from 'terra-icon';
+
+const SvgIcon = customProps => {
+  return (
```
Comment (mhemesath) [inline src/template.txt]: another NIT, if you have a simple arrow function you don't need a return statement if you do this

```js
customProps => (
  <Icon {...customProps}>
    <%= icon.children %>
  </Icon>
);
```

Comment (Matt-Butler) [inline src/template.txt]: Good catch
https://github.com/cerner/terra-react-svg-loader/pull/1/commits/ed651874c04de175cce076751e01be285bac751b

File: lib/template.txt [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
```
Comment (mhemesath) [inline lib/template.txt]: this file looks great!

**Comment by tbiethman** [comment]
+1

**Comment by Matt-Butler** [comment]
Icons worked with terra-button: 
https://github.com/cerner/terra-button/commit/3da82f3fcb2524dbdc3fb15abe1a50574cf325ae

File: tests/jestsetup.js [hunk 7]
```diff
@@ -6,8 +6,10 @@ global.render = render;
 global.mount = mount;
 
 // Skip createElement warnings but fail tests on any other warning
```
Comment (bjankord) [inline tests/jestsetup.js]: What is this needed for? Why do we skip createElement warnings?

File: tests/nodeToJsxNode.test.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import nodeToJsxNode from '../src/nodeToJsxNode';
+
+describe('htmlToReactAttributes', () => {
+  it('attributes should change to jsx', () => {
```
Comment (bjankord) [inline tests/nodeToJsxNode.test.js]: nit: it('hyphenated attributes should change camelCase', () => {

File: tests/nodeToJsxNode.test.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    const input = '<path fill-rule="evenodd" />';
+    document.body.innerHTML = input;
+    expect(nodeToJsxNode(document.body.children[0])).toBe('<path fillRule="evenodd" ></path>');
+    expect(nodeToJsxNode(document.body.children[0])).toMatchSnapshot();
+  });
+  it('attributes should stay the same', () => {
```
Comment (bjankord) [inline tests/nodeToJsxNode.test.js]: nit: it('non hyphenated attributes should remain the same', () => {

File: tests/jestsetup.js (hunk 7 above)
Comment (Matt-Butler) [inline tests/jestsetup.js]: This is from the jest snapshot testing setup:

https://hackernoon.com/testing-react-components-with-jest-and-enzyme-41d592c174f#fda2

File: tests/nodeToJsxNode.test.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    expect(nodeToJsxNode(document.body.children[0])).toBe('<path fill="#232" ></path>');
+    expect(nodeToJsxNode(document.body.children[0])).toMatchSnapshot();
+  });
+  it('attribute class should change to className', () => {
+    const input = '<path class="math" />';
+    document.body.innerHTML = input;
+    expect(nodeToJsxNode(document.body.children[0])).toBe('<path className="math" ></path>');
+    expect(nodeToJsxNode(document.body.children[0])).toMatchSnapshot();
+  });
+  it('it should only return a single element as string', () => {
```
Comment (bjankord) [inline tests/nodeToJsxNode.test.js]: Is the intention that nodeToJXSNode function only accepts a single element/node to transform?
I'm a bit confused by the usage here. Given `<path class="math"/><path class="math"/>` as input to the function, it looks like the function transforms it into one element `<path className="math" ></path>` .

File: tests/nodeToJsxNode.test.js (hunk 8 above)
Comment (Matt-Butler) [inline tests/nodeToJsxNode.test.js]: I wasn't sure how to word this because not all hyphenated attributes change to camelCase.
```
aria-*
data-*
```
are not transformed at all. They remain the same.

File: tests/nodeToJsxNode.test.js (hunk 10 above)
Comment (Matt-Butler) [inline tests/nodeToJsxNode.test.js]: It will only transform and return the first tag. I put this here in case future changes were made with the assumption that it would transform multiple tags.

File: tests/jestsetup.js (hunk 7 above)
Comment (bjankord) [inline tests/jestsetup.js]: Ok, thanks. I left a comment on that post. I'm more just curious about why we need to patch this error message. I'll see if they respond on the linked post. If you remove this code, do the tests still pass?

File: tests/nodeToJsxNode.test.js (hunk 8 above)
Comment (bjankord) [inline tests/nodeToJsxNode.test.js]: Ah, that makes sense. Seeing the test `attributes should change to jsx` followed by `attributes should stay the same`, I was thinking we should make the test names more specific.

Comment (bjankord) [inline tests/nodeToJsxNode.test.js]: Maybe have the first test as is, `attributes should change to jsx' and the second test, `white-listed attributes should stay the same`

Comment (Matt-Butler) [inline tests/nodeToJsxNode.test.js]: Here is the logic on how attributes should be transformed:
https://github.com/cerner/terra-react-svg-loader/pull/1/files#diff-b12ba7751490f455a07c15c294a09cdd

This page describes which attributes need to be changed:
https://facebook.github.io/react/docs/dom-elements.html

File: src/optimizeSvg.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (bjankord) [inline src/optimizeSvg.js]: You might try `// eslint-disable-next-line import/no-extraneous-dependencies` here and remove the comment on line 3.

File: tests/nodeToJsxNode.test.js (hunk 10 above)
Comment (bjankord) [inline tests/nodeToJsxNode.test.js]: If the desired behavior is to only transform and return the first tag I think we should throw an error if more than 1 element is passed in. Then test that case, pass in 2 elements and catch the error.  This would make the function more predictable.

Given `<path class="math"/><path class="math"/>` as input, I find it a bit confusing that 1 element is returned.

In case future changes were made with the assumption that the function would transform multiple tags, we could delete this behavior and add tests around the new functionality.

File: tests/nodeToJsxNode.test.js (hunk 8 above)
Comment (bjankord) [inline tests/nodeToJsxNode.test.js]: Yeah, if you want to update the second test description to `white-listed attributes should stay the same` where white-list refers to the white-list from facebook, I think it adds some clarity. Whatever you want to do here is cool with me.

File: tests/__snapshots__/nodeToJsxNode.test.js.snap [hunk 12]
```diff
@@ -0,0 +1,7 @@
+exports[`htmlToReactAttributes attribute class should change to className 1`] = `"<path className=\"math\" ></path>"`;
```
Comment (bjankord) [inline tests/__snapshots__/nodeToJsxNode.test.js.snap]: I'm not seeing a test for this in `tests/nodeToJsxNode.test.js`. It looks like this functionality is now tested in `tests/htmlToReactAttributes.test.js` and doesn't generate a snapshot. Should this snapshot for className be removed?

File: tests/nodeToJsxNode.test.js (hunk 10 above)
Comment (Matt-Butler) [inline tests/nodeToJsxNode.test.js]: I will remove this test then if it doesn't make sense. This scenario should never occur.

File: tests/icons/guitar.svg [hunk 13]
```diff
@@ -0,0 +1,13 @@
+<svg id="path1" fill-rule="evenodd">
```
Comment (bjankord) [inline tests/icons/guitar.svg]: I'm curious about this file. There are 2 SVG elements in it. The 2nd snapshot test in `tests/transformChildren.test.js` tests this file and returns the children from the first SVG element in this file. The generated snapshot is based on the first SVG element in this file.

Is this the intended behavior for this test?  My guess is the intention was to test the 2nd SVG element, the one with id="guitar". Do we anticipate people to pass in SVG files with multiple SVG elements to loader. This seems like an edge case we might not need to tackle now.

Comment (Matt-Butler) [inline tests/icons/guitar.svg]: Good catch. There should only be one svg per file.

**Comment by bjankord** [comment]
Overall I think this looks good, I have a few comments I'd like to see addressed mainly in the `tests`. I think adding some JSDoc to files in the `src` directory describing their use, input, output would make the code a bit easier to reason about as one is figuring out how all the code works together, but I won't block on that for this PR.

File: tests/jestsetup.js (hunk 7 above)
Comment (Matt-Butler) [inline tests/jestsetup.js]: The tests will still pass if I remove this block. I left it there originally to be consistent with the rest of our repos. I removed it now.

File: tests/__snapshots__/nodeToJsxNode.test.js.snap (hunk 12 above)
Comment (Matt-Butler) [inline tests/__snapshots__/nodeToJsxNode.test.js.snap]: These snapshots are still used. I updated 
htmlToReactAttributes -> nodeToJsxNode

File: tests/jestsetup.js (hunk 7 above)
Comment (bjankord) [inline tests/jestsetup.js]: I'm fine with leaving it, was just curious why it is needed. I'd like to understand why they recommend patching this error message, but I think we should be consistent with the other repos.

File: tests/nodeToJsxNode.test.js (hunk 10 above)
Comment (bjankord) [inline tests/nodeToJsxNode.test.js]: Sounds good

> Yes, the intention is to only pass in a single node.

Do you think it would be valuable to warn if multiple nodes are passed since that is not behavior we intend to support at this time? What is the likelihood multiple nodes would be passed to the function?

File: tests/jestsetup.js (hunk 7 above)
Comment (bjankord) [inline tests/jestsetup.js]: I got a [response](https://medium.com/@bjankord/nice-write-up-artem-sapegin-i-was-curious-what-this-part-was-needed-for-in-the-jestsetup-js-file-72275f8cd334#.a3lu5vco8) back from the author who wrote https://hackernoon.com/testing-react-components-with-jest-and-enzyme-41d592c174f#fda2

I asked about what that code was needed for.

He said, "I actually don’t really remember now and probably it’d be better to remove it from the article."

For now, I think we should leave it in the repo. Though we need to test removing it from the repos which have this and see if the tests still work. That will be easier to test as we move to consolidating the packages under 1 location. If they all still work without it, we should remove it from the generator and update the repos.

File: tests/nodeToJsxNode.test.js (hunk 10 above)
Comment (Matt-Butler) [inline tests/nodeToJsxNode.test.js]: It is extremely unlikely that multiple nodes would ever be passed into this function. I don't think adding a check would be useful. No one should be using the functions in this library directly, only the loader.

File: tests/jestsetup.js (hunk 7 above)
Comment (Matt-Butler) [inline tests/jestsetup.js]: Added it back
